### ISSUE TITLE
LineUp v4: Wrong numerical filter range in side panel for score columns

### ISSUE THREAD

**Issue by thinkh** [datavisyn/tdp_core#352]
* Release number or git hash: lineupjs4 branch with https://github.com/lineupjs/lineupjs/commit/a5db0bd23048a29c9f1817328d2844e7bc68e214
* Web browser version and OS: independent
* Environment (local or deployed): local

### Steps to reproduce

1. Open ranking (e.g., Gene list in Ordino)
2. Add a numerical score column (e.g., Single Cell Line Score in Ordino)

### Observed behavior

The filter in the side panel is wrong. It seems to be the correct histogram in the background, but the filter range is incorrect. Opening the filter dialog shows the same histogram and selects the whole range. The histogram in the column summary looks also correct.

![grafik](https://user-images.githubusercontent.com/5851088/78986572-d96f9480-7b2b-11ea-9e4c-ee7d0d86e282.png)

Checking the response of the score API request and logging the min/max value shows that the values are correct:

![grafik](https://user-images.githubusercontent.com/5851088/78986966-d6c16f00-7b2c-11ea-86fe-506bbe937b6c.png)

### Expected behavior

* The filter in the side panel should match the filter dialog.
* The whole filter range should be selected.

**Comment by thinkh** [comment]
@sgratzl From your first look, could that be a LineUp bug? Otherwise we would start investigating in tdp_core.

**Comment by sgratzl** [comment]
as far as I remember the scoring columns hack in the domain into the column. Please  make sure that still the proper events (as it would in case of a mapping change) will be triggered

**Comment by thinkh** [comment]
Thanks for the hint, we will check this.

**Comment by thinkh** [comment]
I found the root cause after debugging in tdp_core and lineupjs, which was really hard to find.

### Going down the rabbit hole

In order to understand how the error occurs I will quickly recap the steps when loading a score column:

1. Create column description with some default values ([utils.ts](https://github.com/Caleydo/tdp_publicdb/blob/8729fedfb49b8beb1fff72af3936967f288186e1/src/scores/utils.ts#L60-L67))
   * e.g., `domain: [0, 15]` for relative copy number ([config.ts](https://github.com/Caleydo/tdp_publicdb/blob/920738ffb3b08a5274dd5d90e055473ace0f5414/src/config.ts#L290))
   * All score columns (NumberColumn, NumbersColumn, BoleanColumn, ...) extend the `ValueColumn`, which provides the basic implementation for lazy loaded values
   * Lazy columns are marked in tdp_core ([column.ts](https://github.com/datavisyn/tdp_core/blob/7af4919528d8ffadcc0f1638a6b7579d83d7db90/src/lineup/internal/column.ts#L37))
   * The flag is picked up in the `ValueColulmn`: `this.loaded = desc.lazyLoaded !== true;`
   * As long as `loaded` is `false` the loading renderer is shown
2. Call score's `compute()` function which retrieves the data from the API 
3. Once the data loading has finished the success function is called ([column.ts](https://github.com/datavisyn/tdp_core/blob/7af4919528d8ffadcc0f1638a6b7579d83d7db90/src/lineup/internal/column.ts#L63-L72))
   https://github.com/datavisyn/tdp_core/blob/7af4919528d8ffadcc0f1638a6b7579d83d7db90/src/lineup/internal/column.ts#L63-L72
4. `patchColumn()` calculates the `domain` based on the loaded data and sets the domain directly to the column and mapping 
   https://github.com/datavisyn/tdp_core/blob/7af4919528d8ffadcc0f1638a6b7579d83d7db90/src/lineup/internal/column.ts#L106-L119
   * **Very dirty hack** since LineUp is not aware of this change and does not update internally -> we will see in a few steps what will happen
5. `markLoaded()` sets the `loaded` flag of the `ValueColumn` to `true`
   * This in turn triggers a re-rendering ([ValueColumn.ts](https://github.com/lineupjs/lineupjs/blob/89d3e8a2d241e2a10d2a24f0ecc6bb5ae0e2329b/src/model/ValueColumn.ts#L84))
   * The `EVENT_DIRTY_HEADER` is passed via EngineRenderer, SidePanel, SidePanelEntryVis down to the histogram
6. The [`interactiveSummary()`](https://github.com/lineupjs/lineupjs/blob/d962e3cac41c1b7ea40a095aaf208a944f94bbfc/src/renderer/HistogramCellRenderer.ts#L91) in LineUp's HistogramCellRenderer.ts creates the histogram and filter functionality
   ```ts
   const fContext = createFilterContext(col, context);
   ```
7. The `createFilterContext()` stores (a copy) of the `domain` array and caches it
   ```ts
   const domain = col.getMapping().domain; // CACHES THE FIRST DOMAIN (e.g., [0, 15])
   const format = col.getNumberFormat();
   const percent = (v: number) => Math.round(100 * (v - domain[0]) / (domain[1] - domain[0]));
   const unpercent = (v: number) => ((v / 100) * (domain[1] - domain[0]) + domain[0]);
   ```
   * `percent()` is used to calculate the percentage value for the CSS style
8. When [updating the histogram + filter](https://github.com/lineupjs/lineupjs/blob/d962e3cac41c1b7ea40a095aaf208a944f94bbfc/src/renderer/HistogramCellRenderer.ts#L109) a new [`createFilterInfo()`](https://github.com/lineupjs/lineupjs/blob/d962e3cac41c1b7ea40a095aaf208a944f94bbfc/src/renderer/HistogramCellRenderer.ts#L206-L215) is created with the new domain
   ```ts
    function createFilterInfo(col: IMapAbleColumn, filter = col.getFilter()): IFilterInfo<number> {
      const domain = col.getMapping().domain; // HERE THE UPDATED DOMAIN IS USED (e.g., [0.39, 4.41])
      const filterMin = isFinite(filter.min) ? filter.min : domain[0];
      const filterMax = isFinite(filter.max) ? filter.max : domain[1];
      return {
        filterMissing: filter.filterMissing,
        filterMin, // = 0.39
        filterMax // = 4.41
      };
    }
   ```
9. When calculating the percent value for the CSS style the new domain is compared to the cached old domain value
   * The result is the initially shown range filter

The [HistogramCellRenderer.ts](https://github.com/lineupjs/lineupjs/pull/265/files#diff-22c636d3a16b629154da8ac27622d99c) and the corresponding functions were last modified in commit https://github.com/lineupjs/lineupjs/commit/c8e6e4623b9741c9a37ad60749bd88577e6cb536, which belongs to https://github.com/lineupjs/lineupjs/pull/265 and was a major refactoring.

### Possible solutions

#### Avoid caching of the domain

A possible solution would be to avoid the caching of the domain array in the `createFilterContext` and always request the most recent domain mapping:

```ts
  // const domain = col.getMapping().domain;
  const format = col.getNumberFormat();
  const percent = (v: number) => {
    return Math.round(100 * (v - col.getMapping().domain[0]) / (col.getMapping().domain[1] - col.getMapping().domain[0]));
  };
  const unpercent = (v: number) => {
    return ((v / 100) * (col.getMapping().domain[1] - col.getMapping().domain[0]) + col.getMapping().domain[0]);
  };
```

Since the cache was implemented for a reason and there might be other places that I have not found yet, we should aim for a different solution.

#### Check data mapping dialog

Alternatively, we should check when this cached domain is cleared, e.g., when modifying the domain in the data mapping dialog.

**Comment by thinkh** [comment]
It seems to be a general LineUp bug with the domain in the histogram. I filed a bug report at https://github.com/lineupjs/lineupjs/issues/314.

**Comment by thinkh** [comment]
I compiled a lineupjs build with PR https://github.com/lineupjs/lineupjs/pull/315 and tested it again. This time it successfully works for me. I did not need to change anything or add a `column.markDirty()` as suggested.

![tdp_core-domain](https://user-images.githubusercontent.com/5851088/80469755-4f079d00-8941-11ea-87e0-913a4743f937.gif)

**Comment by thinkh** [comment]
I merged PR https://github.com/lineupjs/lineupjs/pull/315.

@oltionchampari can you please try it again on your instance to verify my result? Please make sure that you update `lineupjs` in your node_modules.

**Comment by oltionchampari** [comment]
> I merged PR [lineupjs/lineupjs#315](https://github.com/lineupjs/lineupjs/pull/315).
> 
> @oltionchampari can you please try it again on your instance to verify my result? Please make sure that you update `lineupjs` in your node_modules.

The fix works works as expected for me too.

![number_filter](https://user-images.githubusercontent.com/51322092/80471453-b888ab00-8943-11ea-872d-d3e7895b6a25.gif)

**Comment by thinkh** [comment]
@oltionchampari Thanks for double checking. Then I close this issue, as the fix is included automatically in the next lineupjs4 build.
